content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -889,6 +853,200 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+    agent any
+    environment {
+        EXAMPLE_CREDS = credentials('example-credentials-id')
+    }
+    stages {
+        stage('Example') {
+            steps {
+                /* WRONG! */
+                sh("curl -u ${EXAMPLE_CREDS_USR}:${EXAMPLE_CREDS_PSW} https://example.com/")
+            }
+        }
+    }
+}
+// Script //
+----
+Should Groovy perform the interpolation, the sensitive value will be injected directly into the arguments of the `sh` step, which among other issues, means that the literal value will be visible as an argument to the `sh` process on the agent in OS process listings.
+Using single-quotes instead of double-quotes when referencing these sensitive environment variables prevents this type of leaking.
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+    agent any
+    environment {
+        EXAMPLE_CREDS = credentials('example-credentials-id')
+    }
+    stages {
+        stage('Example') {
+            steps {
+                /* CORRECT */
+                sh('curl -u $EXAMPLE_CREDS_USR:$EXAMPLE_CREDS_PSW https://example.com/')
+            }
+        }
+    }
+}
+// Script //
+----
+
+
+==== Injection via interpolation
+
+[WARNING]
+======
+Groovy string interpolation can inject rogue commands into command interpreters via special characters.
+======
+
+Another note of caution. Using Groovy string interpolation for user-controlled variables with steps that pass their arguments to command interpreters such as the `sh`, `bat`, `powershell`, or `pwsh` steps can result in problems analogous to SQL injection.
+This occurs when a user-controller variable (generally an environment variable, usually a parameter passed to the build) that contains special characters (e.g. `/ \ $ & % ^ > < | ;`) is passed to the `sh`, `bat`, `powershell`, or `pwsh` steps using Groovy interpolation.
+For a simple example:
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent any
+  parameters {
+    string(name: 'STATEMENT', defaultValue: 'hello; ls /', description: 'What should I say?')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        /* WRONG! */
+        sh("echo ${STATEMENT}")
+      }
+    }
+  }
+}
+// Script //
+----
+
+In this example, the argument to the `sh` step is evaluated by Groovy, and `STATEMENT` is interpolated directly into the argument as if `sh('echo hello; ls /')` has been written in the Pipeline.
+When this is processed on the agent, rather than echoing the value `hello; ls /`, it will echo `hello` then proceed to list the entire root directory of the agent.
+Any user able to control a variable interpolated by such a step would be able to make the `sh` step run arbitrary code on the agent.
+To avoid this problem, make sure arguments to steps such as `sh` or `bat` that reference parameters or other user-controller environment variables use single quotes to avoid Groovy interpolation.
```

Comment:
```suggestion
To avoid this problem, make sure arguments to steps such as `sh` or `bat` that reference parameters or other user-controlled environment variables use single quotes to avoid Groovy interpolation.
```